logging/net_appender.go

```diff
@@ -206,40 +207,45 @@ func (nl *NetAppender) backgroundWorker() {
 	}
 }
 
-// Sync will flush the internal buffer of logs.
-func (nl *NetAppender) Sync() error {
-	for {
-		batch := func() []*apppb.LogEntry {
-			nl.toLogMutex.Lock()
-			defer nl.toLogMutex.Unlock()
-
-			if len(nl.toLog) == 0 {
-				return nil
-			}
-
-			batchSize := writeBatchSize
-			if len(nl.toLog) < writeBatchSize {
-				batchSize = len(nl.toLog)
-			}
+// Returns whether there is more work to do or if an error was encountered while trying to ship logs
+// over the network.
+func (nl *NetAppender) syncOnce() (bool, error) {
+	nl.toLogMutex.Lock()
```

Comment:
The big change here that fixes the test is holding the lock for the entire sync call. Before the background worker could get a batch (pause before sending) concurrently with the test calling Sync getting the second batch. But because we used to drop the lock while sending results, there was no guarantee in the order the grpc log requests were sent/received.

Reply:
> Before the background worker could get a batch (pause before sending) concurrently with the test calling Sync getting the second batch

Ah yep good find 🧑‍🔧 .